Refactor Completer API

edzkite opened the pull request:
## References
Resolves jupyterlab/jupyterlab#7044

See https://github.com/jupyterlab/jupyterlab/issues/7044#issuecomment-594249225 for high level architecture.

Changes from architecture spec:
- Rather than transform current items into `ICompletionItems`, there are two flows: the current completer flow and a new flow just for `ICompletionItems`
  - This was done in the spirit of backwards compatibility, and cleaner deprecation if desired
- Burden of deduping and sorting `ICompletionItems` moved to completion provider extensions, to give extensions more control over that behavior.
- Punting implementation of `isIncomplete` / `resolve` for documentation. The story there should probably be to a separate issue / PR with community discussion.


## Code changes

#### Handler
* Added `ICompletionItems` and `ICompletionItem` interfaces to encapsulate completions
* Added `ICompletionItems` as an optional property to preexisting `IReply` interface to keep backwards compatibility with preexisting connectors
* In the response handler, if the `ICompletionItems` property `items` is in the reply, send to the model right away

####  Model
* Create getters and setters for `ICompletionItems`
For completion items:
* Create function `_markup()` that detects prefix matches and adds `<mark>` tags
* Add `findOrderedCompletionItemTypes` to add new types into a set, before sorting as an array to append to `KNOWN_TYPES`

#### Widget:
* Incorporate ICompletionItems in `onUpdateRequest` and `createItemNode`
* Add function `_createCompletionItemNode` to create node for completion item.
* Move creation of legacy completions into `_createIItemNode `

#### Tests:
* Duplicated widget tests that leveraged `setOptions` to use `setCompletionItems` instead

## User-facing changes
N/A

## Backwards-incompatible changes
N/A

saulshanabrook commented:
Just had a quick review, to look for API breaking changes. Just wondering if these are intentional or if you meant to keep it backward compatible.

edzkite commented:
We want to avoid breaking API changes. To achieve that, we keep the `IRequest` and `IReply` interfaces that the `completer-extension` or other extension packages would leverage. Any new interfaces are used internally within the `completer` package.

Are there any other potential breaking changes that I may have missed?

saulshanabrook commented:
Well since we expose the `Completer` model, the changes to the signatures there are breaking changes:

* https://github.com/jupyterlab/jupyterlab/pull/8080/files#diff-a87b8d7967bd482ef72fe8656b494987L174
* https://github.com/jupyterlab/jupyterlab/pull/8080/files#diff-a87b8d7967bd482ef72fe8656b494987L181
* https://github.com/jupyterlab/jupyterlab/pull/8080/files#diff-a87b8d7967bd482ef72fe8656b494987L219


And the compelter widget also changes the signatures of some methods:

* https://github.com/jupyterlab/jupyterlab/pull/8080/files#diff-122cc21cda6dfa0ff1bf6dea9e87f346L616
* https://github.com/jupyterlab/jupyterlab/pull/8080/files#diff-122cc21cda6dfa0ff1bf6dea9e87f346L683
* https://github.com/jupyterlab/jupyterlab/pull/8080/files#diff-122cc21cda6dfa0ff1bf6dea9e87f346R691
* https://github.com/jupyterlab/jupyterlab/pull/8080/files#diff-122cc21cda6dfa0ff1bf6dea9e87f346L731

edzkite commented:
Would only adding new signatures, while keeping the original interface as is, work?

saulshanabrook commented:
> Would only adding new signatures, while keeping the original interface as is, work?

I believe so. If you are around this week for our [dev meeting](https://github.com/jupyterlab/jupyterlab#weekly-dev-meeting) you might want to come then to talk about this PR.

krassowski commented on packages/completer/src/widget.ts:
```diff
@@ -725,30 +698,25 @@ export namespace Completer {
       li.className = ITEM_CLASS;
       // Set the raw, un-marked up value as a data attribute.
-      li.setAttribute('data-value', item.raw);
+      li.setAttribute('data-value', item.insertText || item.label);
```
This will work, but we may want to modify it slightly to make the future snippet support in extensions easier; to achieve that we may need more information from `ICompletionItem` available in the handler (rather than the `insertText` or `label` alone).

So, could we have an identifier (`ICompletionItem.id`) which would be replaced with the actual `ICompletionItem` object in wherever the `_selected` signal is emitted? Then the signal receiver ([`CompletionHandler.onCompletionSelected`](https://github.com/jupyterlab/jupyterlab/blob/2f049d57563b628b2e8d6202d716fb7bb7712687/packages/completer/src/handler.ts#L176)) would accept `item:  ICompletionItem` rather than `val: string`.

Form the extension point of view, the extension would only need to modify or re-implement `CompletionHandler` to get fancy new features such as snippets.

krassowski replied:
On the other hand, if one wants to heavily modify the completer they will need to modify the renderer as well anyways...

edzkite replied:
While this would certainly be a useful change, like pagination of completion items or asynchronous handling of documentation, this would probably be best served as a follow up. What do you think?

krassowski replied:
I agree.

krassowski commented on packages/completer/src/handler.ts:
```diff
@@ -477,6 +511,93 @@ export namespace CompletionHandler {
+    /**
+     * isIncomplete is true if this is a partial list of completion items.
+     */
+    isIncomplete: boolean;
```
What is the use case for this? I do not see `isIncomplete` being accesses anywhere, nor being set to true in any of tests.

krassowski replied:
Ok, I see this discussed in https://github.com/jupyterlab/jupyterlab/issues/7044#issuecomment-594249225 - sorry. Maybe larger parts of the comment could be embedded in the interface documentation?

edzkite replied:
I've opted to remove `isIncomplete` for now. On second thought, this is probably best tackled in a follow up issue / PR.

saulshanabrook commented on packages/completer/src/widget.ts:
```diff
@@ -725,6 +805,45 @@ export namespace Completer {
+        allowedTags: ['mark']
+      });
+
+      if (item.type) {
```
Do you think it would make sense to refactor some of the code here and in `createItemNode` that is common? Os have one call the other?

saulshanabrook commented on packages/completer/src/model.ts:
```diff
@@ -355,6 +385,33 @@ export class CompleterModel implements Completer.IModel {
+    let items = JSONExt.deepCopy(this._completionItems.items);
+    let results: CompletionHandler.ICompletionItem[] = [];
+    for (let item of items) {
+      // See if insert text matches query string
```
```suggestion
      // See if label matches query string
```

saulshanabrook commented on packages/completer/src/model.ts:
```diff
@@ -355,6 +385,33 @@ export class CompleterModel implements Completer.IModel {
+          match.indices,
+          Private.mark
+        );
+        if (!item.insertText) {
+          item.insertText = item.label;
+        }
```
Why do we need this here? I thought in other places it does an || between insertText and label, to choose label if insertText is empty.

saulshanabrook replied:
Ah I see, makes sense, thank you! Could you add a comment here to say that?

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -7,7 +7,12 @@ import { Text } from '@jupyterlab/coreutils';
 
```
Could you add a little docstring somewhere, wherever you think is most appropriate, generally talking about how we had this old completion data structures that worked for the kernel, but now we extended them to provide more hooks for the UI, and if you are a new completion engine, you should use the new ones?

edzkite commented on packages/completer/src/handler.ts:
```diff
@@ -500,6 +588,11 @@ export namespace CompletionHandler {
+    /**
+     * Hook for extensions to send ICompletionItems.
+     */
+    items?: ICompletionItems;
```
While I like the idea in principle, this may require some extensive changes to existing downstream extensions. For instance, [jupyterlab-lsp](https://github.com/krassowski/jupyterlab-lsp/blob/master/packages/jupyterlab-lsp/src/adapters/jupyterlab/components/completion.ts#L32) exclusively uses a `LSPConnector` that inherits from `DataConnector` to communicate with JupyterLab.

Perhaps I could improve the documentation within `IReply` to make it less confusing?

saulshanabrook replied:
> For instance, jupyterlab-lsp exclusively uses a LSPConnector that inherits from DataConnector to communicate with JupyterLab.

It doesn't seem like that big a change for them. @krassowski do you have thoughts on this?

Is it true that if you provide `items` then the `matches` and `metadata` are unused? So a downstream user would have to just create dummy things to pass in there?

krassowski replied:
I'm ok with either (I might slightly prefer @saulshanabrook suggestion). If the current form was to stay, should `matches` and `metadata` get deprecated?

saulshanabrook replied:
>  the fact that other packages also use the IDataConnector interface to fetch requests make me a little hesitant about introducing this new pattern.

I mean you could just use a DataConnector thing for the new param as well. But we only use the `fetch` function in it, so it just seems like an overkill abstraction IMHO but we do use it other places, so I would be fine reusing it here.

edzkite replied:
I added `fetchItems` in `CompletionHandler.IOptions` as well as `ICompletionManager.ICompletableAttributes`, making `connector` optional. If `connector` is not set, I instantiate a `DummyConnector` that rejects the `fetch` call.

saulshanabrook replied:
Great! Thank you!

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -477,6 +486,85 @@ export namespace CompletionHandler {
+   * Implementers of this interface should be responsible for
+   * deduping and sorting the items in the list.
+   */
+  export interface ICompletionItems {
```
> My only concern, with the documentation use case in mind, is that if issues arise requiring additional enhancements, they would require a breaking change, potentially delaying fixes. Happy to remove this if it's an over optimization.

Yeah that's fair... I would say that at the moment the simplicity of removing it would be nice, unless we can imagine how we would extend it now.

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -477,6 +486,85 @@ export namespace CompletionHandler {
+  /**
+   * Completion item object based off of LSP CompletionItem
+   */
+  export interface ICompletionItem extends PartialJSONObject {
```
> In the model, I use a couple of methods from the JSONExt library that require this interface to be a PartialJSONObject

Hm, I don't believe you have to extend it though for it to be compliant. Or you shouldn't have to!

saulshanabrook replied:
Can we just not have it allow any key? And if we need to pass it to a JSONExt thing, just cast it to a `PartialJSONObject`?

edzkite replied:
> Can we just not have it allow any key?
```typescript
export interface ICompletionItem  {
    [key: any]: PartialJSONValue | undefined
```
does not work, since the index signature no longer matches `PartialJSONObject`, which requires keys as strings.

> And if we need to pass it to a JSONExt thing, just cast it to a PartialJSONObject?
Since we're passing in an array to `JSONExt`, we'd need to cast `ICompletionItems` into a `ReadonlyJSONArray`. Unfortunately, this would require the individual items to adhere to `PartialJSONValue` which itself is a type alias that encapsulates `PartialJSONObject`.

At a high level though, is the concern that the interface is too restrictive if we extend from `PartialJSONObject`, or force the keys of `ICompletionItem` to be strings?

Given that we care to use `JSONExt` for its `deepCopy` and deep equality functionality, there may be other libraries we could use apart from Lumino that gives us more type flexibility, if that's an avenue to explore.

saulshanabrook replied:
> At a high level though, is the concern that the interface is too restrictive if we extend 

No the concern is that it's too open ended, the opposite. I think the interface should *only* specify the keys you need from it, and not have any generic keys in it.

What do you need from `JSONExt`?

You could just cast when you pass it in (I know this isn't ideal, but it seems preferably then having an overly permissive public interface).

```typescript
function takesJSON(j: PartialJSONObject): void {

}

interface IMyInterface {
  someKey: string;
}

const i: IMyInterface = {someKey: "hi there"}

takesJSON(i as unknown  as PartialJSONObject);
```

saulshanabrook replied:
To give some more reasoning, my assumption is that if an interface allows arbitrary key/values I assume that those will be *used* in some way by the function that takes that interface. But that isn't the case here. We will only by using the the keys you manually specified. So it is a clearer contract to me with the callers if we specify that by not extending from the PartialJSONObject or allowing arbitrary keys.

edzkite replied:
Thanks for the clarification! Casting to `unknown` first resolved the issues.

saulshanabrook replied:
Thank you! Thanks for bearing with my comments.

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -434,19 +464,51 @@ export class CompletionHandler implements IDisposable {
+    if (!model) {
+      return;
+    }
+    if (reply.items && model.setCompletionItems) {
```
I believe `reply.items` will always be a true? an empty array is truthy.

```suggestion
    if (model.setCompletionItems) {
```

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -474,7 +536,82 @@ export namespace CompletionHandler {
+
+  /**
+   * Completion item object based off of LSP CompletionItem.
+   * Compared to the old kernel completions interface, this enhances the completions UI to support:
```
Thank you! This is perfect!

edzkite commented:
Awesome, thanks @saulshanabrook!

afshin commented:
I have a few thoughts on this, sorry it took so long for me to respond.

- Thank you for working so hard to keep backward compatibility. I know it's difficult to do, but it's a big deal.
- It seems like there should be a way to modify the data connector's `fetch()` signature to return`CompletionHandler.ICompletionIReply | CompletionHandler.ICompletionItemsReply` and retain backward compatibility. It seems to me like the `CompletionHandler` should only make one kind of fetch and process one kind of reply. We can check within that reply what the actual shape of the returned object is and handle it there while maintaining backward compatibility.
- Your idea that the deduping logic should be the responsibility of the data provider is a good one, so perhaps the default connector should be doing that deduping and any client whose connector wants to replace the default one can do that as well.
- I would remove the concept of `type ICompletionItems` and instead just use `CompletionHandler.ICompletionItem[]`.

edzkite commented:
@afshin Thanks for the feedback! I think your proposals make a lot of sense, especially with `fetch`.

The biggest pain point I've had with registering `fetchItems` directly is that if you have some state that you want to manage to help with fetching items, implemented as class, you have to make sure to bind `this` to the class before registering, whereas previously, you just pass in the connector and you're good to go. I could see it being a potential gotcha for extension developers.

@saulshanabrook What do you think?

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -474,7 +536,82 @@ export namespace CompletionHandler {
      * rejected promises.
      */
-    connector: IDataConnector<IReply, void, IRequest>;
+    connector?: IDataConnector<IReply, void, IRequest>;
```
@afshin and I had a good discussion about this in the meeting today. My takeaway was that we should allow you to specify either type of connector, and then after you call fetch, check the return type to see which type was provided, by looking at the keys provided in the object 🤷.

Maybe with the idea that we would deprecate the first option eventually? 

```suggestion
    connector: IDataConnector<IReply, void, IRequest> | IDataConnector<ICompletionItemsReply, void, IRequest>;;
```

edzkite replied:
I like it! I'll go ahead and make this change.

edzkite replied:
On this topic, I'm using type assertions to differentiate the different replies, but it seems kind of messy. Do you guys have any guidance on the best practice here? ([Original implementation](https://github.com/jupyterlab/jupyterlab/blob/380ad1a4ce8ba0f5a7b0490f3119f498cae776ae/packages/completer/src/handler.ts#L359) for reference)

```typescript
    const response = this._connector.fetch(request);
    return (response as Promise<
      CompletionHandler.ICompletionItemsReply | undefined
    >)
      .then(reply => {
      // ...
        if (!reply.items) {
          throw new Error(
            'Casting response as Promise<CompletionHandler.IReply'
          );
        }

        this._onFetchItemsReply(state, reply);
      })
      .catch(_ => {
        (response as Promise<CompletionHandler.IReply | undefined>)
          .then(reply => {
               //...
               this._onReply(state, reply);
          })
          .catch(_ => this._onFailure());
      });
```

saulshanabrook replied:
Yeah it's messy... But I think you are doing it right. If you used the tag union approach like you mentioned below it might be a bit cleaner?

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -59,6 +61,9 @@ export class CompletionHandler implements IDisposable {
     void,
```
I am pretty torn up about what to do with this existing getter and setter.

Do we know if anyone is using them?

I am tempted to just add a comment that they are both deprecated, keep the signatures the same for now, returning a dummy connector in the getter. 

Then remove them in 3.0?

edzkite replied:
I think the only instance of it being used in core is [here](https://github.com/jupyterlab/jupyterlab/pull/8080/files#diff-472ee5441f1ad3f53f4e4a22d97bf9dbR367).

I don't believe it's being used by the LSP extension.

edzkite replied:
What if we used [tagged unions](https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions) by adding an optional property in `IDataConnector`? 

This way, new connectors can identify themselves, either via boolean or string identifier. If the identifier is present, we return the dummy connector, otherwise we return `this._connector` casted to keep the getter as is.

saulshanabrook replied:
That would be nice!

edzkite commented:
I updated the implementation so that `Connector` can now support responses of both `IReply` and `ICompletionItemsReply`.

To assist in determining which connector is being used, I added an optional tag to `IDataConnector` and exported a constant named `ICompletionItemsResponseType` that extensions wishing to use the new API can set. 

To preserve the getter signature, I changed the implementation such that if the identifier is present, we return the dummy connector, otherwise we return this._connector casted to the preexisting value.

Let me know what you guys think!

saulshanabrook commented on packages/statedb/src/interfaces.ts:
```diff
@@ -29,6 +29,13 @@ import { ISignal } from '@lumino/signaling';
+   * Used as a discriminant if a property has multiple connectors
+   * with different response types.
+   */
+  responseType?: string;
```
My preference I think is this to not go in the main IDataConnector type, but keeping it localized to the completer code. I added a comment up there to suggest how it might be done?

saulshanabrook commented on packages/completer/src/tokens.ts:
```diff
@@ -49,20 +49,19 @@ export namespace ICompletionManager {
+          void,
+          CompletionHandler.IRequest
+        >
+      | IDataConnector<
+          CompletionHandler.ICompletionItemsReply,
+          void,
+          CompletionHandler.IRequest
+        >;
```
What if this added the new field so we could differentiate it? Maybe like this?

```suggestion
      | (IDataConnector<
          CompletionHandler.ICompletionItemsReply,
          void,
          CompletionHandler.IRequest
        > & {responseType: typeof CompletionHandler.ICompletionItemsResponseType});
```

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -616,6 +629,8 @@ export namespace CompletionHandler {
     items: CompletionHandler.ICompletionItems;
   }
 
+  export const ICompletionItemsResponseType = 'ICompletionItemsReply';
```
If we export this as a const type, then I believe when we do `typeof` below it will only allow this particular string.

```suggestion
  export const ICompletionItemsResponseType = 'ICompletionItemsReply' as const;
```

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -434,19 +481,48 @@ export class CompletionHandler implements IDisposable {
+  private _connector:
+    | IDataConnector<CompletionHandler.IReply, void, CompletionHandler.IRequest>
+    | (IDataConnector<
+        CompletionHandler.ICompletionItemsReply,
+        void,
+        CompletionHandler.IRequest
+      > & {
+        responseType: typeof CompletionHandler.ICompletionItemsResponseType;
+      });
```
We should probably make this a named type since we use it a few places here and in the token.

saulshanabrook commented on packages/completer/src/handler.ts:
```diff
@@ -349,34 +359,76 @@ export class CompletionHandler implements IDisposable {
     const request: CompletionHandler.IRequest = { text, offset };
 
-    return this._connector
+    if ('responseType' in this._connector) {
+      return (this._connector as IDataConnector<
+        CompletionHandler.ICompletionItemsReply,
+        void,
+        CompletionHandler.IRequest
+      >)
+        .fetch(request)
```
Would this work as a [user defined type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards)? Hopefully that would let us avoid the `as` b/c it would just infer the more specific type on this branch.

edzkite commented:
The merge conflicts against master for this PR, as well as https://github.com/jupyterlab/jupyterlab/pull/8137, https://github.com/jupyterlab/jupyterlab/pull/8140, and https://github.com/jupyterlab/jupyterlab/pull/8141 should all be resolved.

There may be some conflicts that arise between the UI PRs. Let me know if there's anything I can do to help in the process. Thanks!

krassowski commented:
@saulshanabrook apologies for a silly question, but did this one got included in 2.2? I do not see it in the changelog...

bollwyvl commented:
Just looking at the first file that changed, it would appear this did land in 2.2.0:

https://github.com/jupyterlab/jupyterlab/blob/v2.2.0/packages/completer-extension/src/index.ts#L96-L99

saulshanabrook commented:
Yeah I believe it did land in that release, I probably just missed it when generating the changelog! Thanks for catching that, we should add it to the changelog in master for 2.2.0.
